### PR TITLE
add authentication, user model, crud on users

### PR DESCRIPTION
**Pull request by GZardiashvili**

### CODE REVIEW

File: backend/auth/helpers/helper.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const bcrypt = require('bcrypt');
+const User = require('../../models/user');
+
+async function getUserByEmail(email) {
+  const user = await User.findOne({ email: new RegExp(email, 'i') }).exec();
+
+  return user;
```
Comment (Lokki17) [inline backend/auth/helpers/helper.js:7]: return User.findOne({ email: new RegExp(email, 'i') }).exec()

File: backend/auth/login.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+const express = require('express');
+require('dotenv').config();
+const jwt = require('jsonwebtoken');
+const helper = require('./helpers/helper');
+
+const router = express.Router();
+
+router.post('/', async (req, res) => {
+  const user = await helper.loginUser(req.body.email, req.body.password);
+  if (user) {
```
Comment (Lokki17) [inline backend/auth/login.js:10]: move logic to service layer

File: backend/auth/passport.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+const helper = require('./helpers/helper');
+
+const jwtCallback = async (jwt_payload, done) => {
+  const user = await helper.getUserByEmail(jwt_payload.email);
+
+  if (user) {
+    return done(null, user);
+  }
+
+  return done(null, false);
```
Comment (Lokki17) [inline backend/auth/passport.js:10]: return done(null, user) and return done(null, false)
false is not opposit to user (object)

File: backend/controllers/register.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+const express = require('express');
+const bcrypt = require('bcrypt');
+const User = require('../models/user');
+
+const router = express.Router();
+
+router.post('/', async (req, res) => {
+  const body = req.body;
+
+  const user = new User({
```
Comment (Lokki17) [inline backend/controllers/register.js:10]: move logic to service layer

File: backend/controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
+const express = require('express');
+const User = require('../models/user');
+
+const router = express.Router();
+
+router.get('/:id?', (req, res) => {
+  if (req.params.id) {
```
Comment (Lokki17) [inline backend/controllers/users.js:7]: User different endpoints to get all and get by id

File: backend/controllers/users.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      .catch((error) => {
+        console.error('The Promise is rejected!', error);
+      });
+  }
+});
+
+router.put('/:id', (req, res) => {
+  const body = req.body;
+
+  const user = {
```
Comment (Lokki17) [inline backend/controllers/users.js:37]: move

File: backend/server.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+require('dotenv').config();
+const mongoose = require('mongoose');
+
+const app = require('./app');
+
+const url = process.env.MONGODB_URI;
+
+mongoose.connect(url);
+
+app.listen(process.env.PORT);
```
Comment (Lokki17) [inline backend/server.js:10]: What if process.env.PORT not set?

File: backend/models/user.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  state: { type: String },
+  city: { type: String },
+  zip: { type: String },
+  birthDate: { type: String },
+  createdDate: { type: Date, default: Date.now },
+  updatedDate: { type: Date, default: Date.now },
+});
+
+userSchema.set('toJSON', {
+  transform: (document, returnedObject) => {
```
Comment (Lokki17) [inline backend/models/user.js:21]: Do you realy need this?

**Review by Lokki17 — CHANGES_REQUESTED** [review]
Fix remarks

File: backend/server.js (hunk 7 above)
Comment (GZardiashvili) [reply backend/server.js:10]: I've heard that it's better to get error in this case, then to have alternate port, but I'll add || 3000

File: backend/auth/services/authService.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+    const token = jwt.sign(payload, process.env.JWT_SECRET, {
+      expiresIn: process.env.JWT_EXPIRES_IN,
+    });
+
+    return {
+      id: user.id,
+      email: user.email,
+      role: user.role,
+      token: `Bearer ${token}`,
```
Comment (Lokki17) [inline backend/auth/services/authService.js:27]: I dont like when you add 'Bearer ' in token. Frontend should do it

File: backend/middlewares/error-handler.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+function errorHandler(err, req, res) {
+  if (err.name === 'UnauthorizedError') {
```
Comment (Lokki17) [inline backend/middlewares/error-handler.js:2]: I'd like to send exception on FE

File: backend/models/user.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  password: { type: String, required: true },
+  role: { type: String, required: true },
+  gender: { type: String },
+  phone: { type: String },
+  country: { type: String },
+  state: { type: String },
+  city: { type: String },
+  zip: { type: String },
+  birthDate: { type: String },
+  createdDate: { type: Date, default: Date.now },
```
Comment (Lokki17) [inline backend/models/user.js:16]: Add 'timestamps' as parameter in Schema and mongoose will add createdAt and updatedAt fields automatically.
```
new Schema(
{....},
  {
    timestamps: true,
  },
)
```

**Review by Lokki17 — CHANGES_REQUESTED** [review]
fix and merge

**Pull request merged**